[BugFix] Unconsistent delvec after disk disable (#41893)

srlch opened the pull request:
Problem:
When we disable a disk, the tablets on it will be clone from the other replica into a new disk. Some abnormal case will happen as following:

1. Begin to disable a disk, drop the tablet from the map in tabletManager. But this step will not clear the TabletUpdates Meta at all, including the delvec cache

2. Before disk disable, suppose we have the following version information: for tablet 12345:
Segment 1 (generated from version 1): k1->1, k2->1, k3->1 (with delvec version2 which delete k1->1) rowset segment id = 0
Segment 2 (generated from version 2): k1->2, rowset segment id = 1
Segment 3 (generated from version 2.1): k1->2, k2->1, k3->1 rowset segment id = 2

3. Suppose we begin the clone process, we want to clone the version 2 from other replica into a new disk. We will clone the version2.1 for this tablet and get Segment 3 (generated from version 2.1): k1->2, k2->1, k3->1 now.

4. But the problem here is that: the snapshot for version2.1 will reset the rowset segment id. It means that the Segment in clone tablet with rowset segment id = 0. Therefore, the incoming apply process for this tablet using tsid = {12345, 0} to acquire the latest delvec. But we acquire it first in the cache which mean that we get the delvec for Segment 1 instead.

Currently, the migration task for primary key table has the similar problem.

Solution:
1. Clear the delvec/dcg cache in _load_from_pb to make sure the new tablet will not affect by the Potentially inconsistent cache
2. Double check before primary key table compaction to ensure that we compact in the right datadir
3. make sure set the tablet state as shut down when drop it from tablet_map

Fixes #41893

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

decster commented on be/src/storage/tablet_updates.cpp:
```diff
@@ -349,10 +349,19 @@ Status TabletUpdates::_load_from_pb(const TabletUpdatesPB& tablet_updates_pb) {
     }
     del_vector_cardinality_by_rssid.clear();
 
+    // make sure there is no delvec/dcg cache for all rowsets in this TabletUpdates
+    // to avoid the potential unconsistents metadata read.
+    for (auto& [id, rowset] : _rowsets) {
+        _clear_rowset_del_vec_cache(*rowset);
+        _clear_rowset_delta_column_group_cache(*rowset);
+    }
+
     l2.unlock(); // _rowsets_lock
 
     RETURN_IF_ERROR(_load_pending_rowsets());
     StorageEngine::instance()->update_manager()->clear_cached_del_vec(tsids);
```
looks like del_vec_cache is cleared multiple times, can we simply this just remove once

srlch replied:
Updated

srlch committed:
[BugFix] Unconsistent delvec after disk disable (#41893)

Problem:
When we disable a disk, the tablets on it will be clone
from the other replica into a new disk. Some abnormal
case will happen as following:

1. Begin to disable a disk, drop the tablet from the map
in tabletManager. But this step will not clear the TabletUpdates
Meta at all, including the delvec cache

2. Before disk disable, suppose we have the following version information:
for tablet 12345:
Segment 1 (generated from version 1): k1->1, k2->1, k3->1 (with delvec version2
                                                            which delete k1->1)
                                                            rowset segment id = 0
Segment 2 (generated from version 2): k1->2, rowset segment id = 1
Segment 3 (generated from version 2.1): k1->2, k2->1, k3->1 rowset segment id = 2

3. Suppose we begin the clone process, we want to clone the version 2 from other
replica into a new disk. We will clone the version2.1 for this tablet and generated
Segment 3 (generated from version 2.1): k1->2, k2->1, k3->1 now.

4. But the problem here is that: the snapshot for version2.1 will reset the rowset segment id.
It means that the clone tablet with rowset segment id = 0. Therefore, the incoming apply process
for this tablet using tsid = {12345, 0} to acquire the latest delvec. But we acquire it first
[linzichao@sandbox-dw starrocks]$ git log
commit 437cd43535975c3890211df59e6004bddf71321b (HEAD -> bugfix_delvec_inconsistent, origin/bugfix_delvec_inconsistent)
Author: srlch <linzichao@starrocks.com>
Date:   Sun Mar 3 15:20:42 2024 +0800

[BugFix] Unconsistent delvec after disk disable (#41893)

Problem:
When we disable a disk, the tablets on it will be clone
from the other replica into a new disk. Some abnormal
case will happen as following:

1. Begin to disable a disk, drop the tablet from the map
in tabletManager. But this step will not clear the TabletUpdates
Meta at all, including the delvec cache

2. Before disk disable, suppose we have the following version information:
for tablet 12345:
Segment 1 (generated from version 1): k1->1, k2->1, k3->1 (with delvec version2
                                                            which delete k1->1)
                                                            rowset segment id = 0
Segment 2 (generated from version 2): k1->2, rowset segment id = 1
Segment 3 (generated from version 2.1): k1->2, k2->1, k3->1 rowset segment id = 2

3. Suppose we begin the clone process, we want to clone the version 2 from other
replica into a new disk. We will clone the version2.1 for this tablet and generated
Segment 3 (generated from version 2.1): k1->2, k2->1, k3->1 now.

4. But the problem here is that: the snapshot for version2.1 will reset the rowset segment id.
It means that the clone tablet with rowset segment id = 0. Therefore, the incoming apply process
for this tablet using tsid = {12345, 0} to acquire the latest delvec. But we acquire it first
in the cache which mean that we get the delvec for Segment 1 instead.

Currently, the migration task for primary key table has the similar problem.

Solution:
1. Clear the delvec/dcg/primary key cache in _load_from_pb to make sure the new tablet will not
affect by the Potentially inconsistent cache
2. set tablet as shut_down in drop_tablets_on_error_root_path
3. Double check before primary key table compaction to ensure that we compaction in the right

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
Add be ut

Signed-off-by: srlch <linzichao@starrocks.com>
